### ISSUE TITLE
[roadmap/tracker] Low precision MoE training

### ISSUE THREAD

**Issue by danielvegamyhre** [pytorch/ao#2147]
Creating this issue as a roadmap/tracker for enabling float8 training for MoEs with token-choice routing. Both core requirements as well as ideas for additional performance optimizations are included.

**UPDATE** 07/22/2025: revised priorities to reflect shifting focus to prioritize mxfp8

**This is not an exhaustive list, but highlights some primary milestones / requirements**

## Compute
- [x] mxpf8  
    - [x] mxfp8 scaled grouped gemm
        - [X] 2d-3d gemm for output and dX (#2848)
        - [x] 2d-2d gemm for dW (https://github.com/pytorch/FBGEMM/pull/4816)
    - [x] torchao differentiable _scaled_grouped_mm support for mxpf8 recipe for dynamic quant before grouped GEMMs
    - [x] triton kernels for per token group scale conversion to blocked swizzled format
        - [x] for 2d inputs with groups along M (#2886)
        - [x] for 3d expert weights (#2894)
        - [x] for 2d in puts with groups along K (https://github.com/pytorch/ao/pull/2956)
- [ ] fp8 rowwise
    - [x] Add torch._scaled_grouped_mm kernel in core 
        - https://github.com/pytorch/pytorch/pull/150374 (by @ngimel)
    - [x] Add differentiable scaled grouped mm with dynamic float8 rowwise quant in torchao
        - https://github.com/pytorch/ao/pull/1969
    - [x] Add custom kernels in torchao for performing per-group scaling on device, to avoid host-device sync
        - https://github.com/pytorch/ao/pull/2064
        - https://github.com/pytorch/ao/pull/2077
    - [ ] Faster inductor codegen kernels for dynamic quant of 3d tensors along dim1: https://github.com/pytorch/pytorch/issues/159769
        - [x] alternatively, handwritten triton kernel faster than torch.compile for this (https://github.com/pytorch/ao/pull/2696)
            - [ ] this also needs to be faster https://github.com/pytorch/ao/issues/2880  
- [ ] fp8 blockwise
    - [ ] quant primitives
    - [ ] Explore DeepGEMM for fp8 blockwise grouped GEMM
    - [ ] triton kernels to do scaling per group without d2h sync

## Communication
I looked at traces and validated "all to all dispatch and shuffle -> grouped gemm -> all to all combine and unshuffle" are all sequentially dependent, so in theory faster/low precision comms should improve performance. There is some overlap with the shared expert computation, but it is not 100% overlap, so there is room for optimization. This will be especially important if/when "all to all" spans multiple nodes, where inter-node network bandwidth is lower than the intra-node NVLink bandwidth. 

- [ ] mxfp8
    - [x] 1d on device all_to_all_v comms kernel (differentiable; with dynamic quant) https://github.com/pytorch/ao/pull/3048
    - [ ]  NVSHMEM put/get primitives instead of `tl.load`/`tl.store` to support inter-node a2a
- [ ] float8 blockwise (P0)

 
## Torchao UX
- [X] Add tensor subclass (ScaledGroupedMMTensor) with an op override for `torch.aten._grouped_mm` => runs differentiable scaled grouped mm
    - https://github.com/pytorch/ao/pull/2275
- [X] Add one line model conversion API, should recursively swap nn.Parameter data tensors of the expert weights with ScaledGroupedMMTensor. 
    - https://github.com/pytorch/ao/pull/2275
- [X] support configurable recipe (fp8 blockwise/rowwise, mxpf8) 

## Compile support
- [x] Compile support for `torch._grouped_mm`
    - done by @bdhirsh in https://github.com/pytorch/pytorch/pull/153384
- [X] Differentiable _scaled_grouped_mm can compile with `fullgraph=True`
- [X] E2E compilation of each TranformerBlock in torchtitan after MoE conversion via tensor subclass approach (fullgraph=False)
- [ ] E2E compilation of each TranformerBlock in torchtitan after MoE conversion via tensor subclass approach (fullgraph=True)

## Distributed support
- [x] Composability with FSDP2 (will likely need something like [this](https://github.com/pytorch/ao/blob/1017c7e3bfe7300a14ed81fa36038684b168b633/torchao/float8/fsdp_utils.py#L129) for the new tensor subclass)
    - [x] mxfp8 (P0)
    - [ ] float8 blockwise (P0)
    - [x] float8 rowwise (P1) https://github.com/pytorch/ao/pull/2413 
- [ ] Composability with TP
    - [x] mxfp8 (P0)
    - [ ] float8 blockwise (P0)
    - [x] float8 rowwise (P1) https://github.com/pytorch/ao/pull/2473
- [ ] Composability with FSDP + TP
    - [x] mxfp8 (P0)
    - [ ] float8 blockwise (P0)
    - [x] float8 rowwise (P1) https://github.com/pytorch/ao/pull/2475
- [ ] Composability with dp2ep as implemented here: https://github.com/pytorch/torchtitan/pull/1324
    - [x] mxfp8 (P0)
    - [ ] float8 blockwise (P0)
    - [x] float8 rowwise (P1) https://github.com/pytorch/ao/pull/2481

**Comment by danielvegamyhre** [comment]
cc @tianyu-l @ngimel @drisspg if you have any feedback on the direction/ideas proposed here

**Comment by yiakwy-xpu-ml-framework-team** [comment]
Hi @danielvegamyhre have your team added quantization tools to convert deepseek bf16 (sft version) to fp8 with block_wise scalars? (better to be aligned with deepseek quantization format Dyanmic FP8 quant and config.json, then I can use it in SGLang and vLLM directly)

I try to seek help to llmcompressor, but they are not ready. Wish updates from torch trainning team if this feature has been supported.

**Comment by danielvegamyhre** [comment]
>Hi @danielvegamyhre have your team added quantization tools to convert deepseek bf16 (sft version) to fp8 with block_wise scalars?

@yiakwy-xpu-ml-framework-team we are currently planning to focus on mxfp8 next due to high interest from various users we've talked to, but fp8 blockwise is on our radar as well! We have an inference only deepseek style blockwise fp8 [prototype](https://github.com/pytorch/ao/tree/main/torchao/prototype/blockwise_fp8). If you want to work on a PR to support training, we'd happily work with you on reviewing and merging that. Otherwise, we'll work on it as soon as we can :)

cc @vkuzo @lessw2020

**Comment by danielvegamyhre** [comment]
Added https://github.com/pytorch/pytorch/issues/159769 to track inductor codegen improvements we'll need for fp8 rowwise backward. Hoping after mxfp8 and fp8 blockwise we'll have time to get fp8 rowwise in good shape as well for PTC.
